osu.Game/Rulesets/Objects/HitObject.cs

```diff
@@ -223,8 +223,16 @@ public IList<HitSampleInfo> CreateSlidingSamples()
         /// <returns>A populated <see cref="HitSampleInfo"/>.</returns>
         public HitSampleInfo CreateHitSampleInfo(string sampleName = HitSampleInfo.HIT_NORMAL)
         {
-            if (Samples.FirstOrDefault(s => s.Name == HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingSample)
-                return existingSample.With(newName: sampleName);
+            // As per stable, all non-normal "addition" samples should use the same bank.
+            if (sampleName != HitSampleInfo.HIT_NORMAL)
+            {
+                if (Samples.FirstOrDefault(s => s.Name != HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingAddition)
+                    return existingAddition.With(newName: sampleName);
+            }
```

Comment:
This change was made by @peppy 

It seems to be a code duplication cleanup for logic that determines the correct bank to apply to a new hitsample. For example if you are making a new hitwhistle sample which is an 'addition', then this code tries to find an existing addition sample to take the bank name of.

In stable, all addition samples have the same bank, because you can only set one bank for all addition samples. This doesn't have to be true in lazer, but currently lazer behaves the same as stable for this in that you can only set one bank for all addition samples.

Reply:
Duplication from where though? If it was deduplication I'd expect the same chunk of code to be gone from somewhere else. But I haven't seen any of that in this diff.

Reply:
Read https://github.com/ppy/osu/pull/23443#issuecomment-1567761667 for the reasoning.
I think this PR introduced the need for the chunk of code that was duplicated and subsequently deduplicated, that's why you don't see it in the final diff.

Reply:
Well the thing is that I'm not confident that it's a safe change to make. Yes `CreateHitSampleInfo()` is frequently used in editor contexts, but it also is frequently used in *non*-editor context. Such as:

https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L94
https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Taiko/UI/DrumSampleTriggerSource.cs#L43-L52

So as far as I can tell this probably changes sample playback behaviour for some objects. So I'd like to know whether this was a change made specifically for the editor flows, or something that is supposed to be valid for everything everywhere.

@peppy mind providing context / commentary on this when able?

Reply:
I did some analysis:

It seems the change allows SpinnerBonusTick to receive an addition bank if the spinner has additions.
This is probably incorrect because spinnerbonus-max and spinnerspin seem to both always use the primary bank.

For taiko Hit and StrongableHitObject, this change allows clap and finish samples added by the Type and IsStrong bindables to use the addition bank.
I think this change is desirable in this case because the user is able to change the addition bank (since this PR), so probably expects all the whistle, clap, and finish samples to have the same addition bank. It would be weird if for example the user sets the clap sample to the normal bank, sets IsStrong to true, and finds a finish sample with the soft bank from the primary bank.

For DrumSampleTriggerSource, this change makes it possible to play the rim and centre samples on different banks if the most valid object has different primary and addition banks. It adds a little bit of mapper intention but maybe we don't want this.

I'm not familiar with taiko hitsounding, but it seems like taiko always plays a single sample per hit so having two different banks on the same object might be a silly idea. If that's the case then we should probably go back and always take the bank of the hitnormal in taiko.

Reply:
So I looked into this myself a bit further and well.....

Here's a test map that I did: [Moderat - Rusty Nails - Copy.zip](https://github.com/user-attachments/files/15817315/Moderat.-.Rusty.Nails.-.Copy.zip)

The thing to note here is that I had all of the sample files just be a speech synthesizer that says the name of the bank. As such:

- For the spinner thing, now I believe it to be an irrelevant concern because stable did not support banks for `spinnerbonus` at all. So the fact this even works is a lazer-only thing and not really advertised as supported anywhere, so I'm fine with leaving it as "undefined behaviour".
- For the taiko thing... well let's just say neither `master` nor this PR follow what stable does.
  - On stable, I can't get the drum to produce anything audible other than "normal". Which means stable does not appear to care about sample bank on individual objects *at all*. The only thing that appears to have an effect there is the thing "default sample settings" (it'll be all normal, all soft, or all drum, always).
  - On `master`, every object with additions will follow its "primary" / `hitnormal` bank,
  - while on this branch, every object with additions will follow its "addition" bank.

I'm not sure what to think of the taiko thing. I'll go ask the taiko NAT maybe for some advice because I'm stupid and know nothing about mapping, but I'm not sure this needs to be a blocking concern for this PR.

With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

Reply:
> I'm not sure what to think of the taiko thing.

As far as I've experienced, taiko players they expect to hear a taiko drum sound and hitsounds either take a backseat being "whatever" or (as we did in lazer) are more present to the point of annoying them.

> With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

Sounds fine.